File: 3235-minimum-cost-to-convert-string-i/3235-minimum-cost-to-convert-string-i.py
class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        # Initialize dist array with infinity so that non-existent edge weights become infinity
        # BE CAREFUL WITH USING * operator, i.e, dist = [[float("inf")] * n] * n
        # You are referencing the same [[float("inf")] * n] n times essentially
        dist = [ [float("inf") for _ in range(26)] for _ in range(26) ]

        # Existing edges
        # Conversion:
        # ord('a') - 97 = 0
        # chr(0 + 97) = 'a'
        for i in range(len(original)):  # == len(changed) == len(cost)
            u = ord(original[i]) - 97
            v = ord(changed[i]) - 97
            w = cost[i]
            print(f"u: {u} v: {v} w: {w}")
            # Same pair in original and changed can appear wiht diff cost,
            # take the smaller
            dist[u][v] = min(dist[u][v], w)

        # itself
        for i in range(26):
            dist[i][i] = 0

        # Find the shortest path for 
        for k in range(26):
            for i in range(26):
                for j in range(26):
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

        for i in dist:
            print('\t'.join(map(str, i)))

        total = 0
        for i in range(len(source)):  # == len(target)
            cost = dist[ord(source[i]) - 97][ord(target[i]) - 97]
            if (cost >= float("inf")):
                return -1
            total += cost

        return total

File: 3235-minimum-cost-to-convert-string-i/3235-minimum-cost-to-convert-string-i.cpp
class Solution {
public:
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        int v = 26; // number of vertices
        int e = original.size(); // number of edges

        // builds a 26x26 adjacency matrix
        vector<vector<long>> dist(v, vector<long>(v, INT_MAX));
        for (int i = 0; i < e; i++) {
            int o = original[i] - 97;
            int c = changed[i] - 97;
            // Edge case 1: "a -> b costs 2" ... "a -> b costs 3" -> keep the minimum cost
            if (dist[o][c] > cost[i]) {
                dist[o][c] = cost[i];
            }
        }

        // Edge case 2: "a -> a costs 100" -> set cost of converting the same char (i.e., not converting) to 0
        for (int i = 0; i < v; i++) {
            for (int j = 0; j < v; j++) {
                if (i == j) dist[i][j] = 0;
            }
        }

        // Floyd=Warshall
        for (int k = 0; k < v; k++) {
            for (int i = 0; i < v; i++) {
                for (int j = 0; j < v; j++) {
                    if (dist[i][j] > dist[i][k] + dist[k][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                    }
                }
            }
        }

        // Iterate through char in source and sum the cost of conversion
        long long total = 0;
        for (int i = 0; i < source.size(); i++) {
            int s = source[i] - 97;
            int t = target[i] - 97;
            // not convertible
            if (dist[s][t] == INT_MAX) return -1;
            total += dist[s][t];
        }

        return total;
    }
};